cookbooks/anisotropic_viscosity/av_material.cc

```diff
@@ -0,0 +1,858 @@
+                                                      .velocities].symmetric_gradient(i, q);
+                        scratch.phi_p[i_stokes] = scratch.finite_element_values[introspection
+                                                                                .extractors.pressure].value(i, q);
+                        ++i_stokes;
+                    }
+                    ++i;
+                }
+                
+                const double eta = scratch.material_model_outputs.viscosities[q];
+                const double one_over_eta = 1. / eta;
+                
+                const bool use_tensor = (anisotropic_viscosity != nullptr);
+                
+                const SymmetricTensor<4, dim> &stress_strain_director = (use_tensor)
+                ?
+                anisotropic_viscosity->stress_strain_directors[q]
+                :
+                dealii::identity_tensor<dim>();
+                
+                
+                
+                const double JxW = scratch.finite_element_values.JxW(q);
+                
+                for (unsigned int i = 0; i < stokes_dofs_per_cell; ++i)
+                    for (unsigned int j = 0; j < stokes_dofs_per_cell; ++j)
+                        if (scratch.dof_component_indices[i] ==
+                            scratch.dof_component_indices[j])
+                            data.local_matrix(i, j) += ((
+                                                         use_tensor ?
+                                                         2.0 * eta * (scratch.grads_phi_u[i]
+                                                                      * stress_strain_director
+                                                                      * scratch.grads_phi_u[j]) :
+                                                         2.0 * eta * (scratch.grads_phi_u[i]
+                                                                      * scratch.grads_phi_u[j]))
+                                                        + one_over_eta * pressure_scaling
+                                                        * pressure_scaling
+                                                        * (scratch.phi_p[i]
+                                                           * scratch.phi_p[j]))
+                            * JxW;
+            }
+        }
+        
+        
+        
+        template <int dim>
+        void
+        StokesPreconditionerAnisotropicViscosity<dim>::
+        create_additional_material_model_outputs(MaterialModel::MaterialModelOutputs<dim> &outputs) const
+        {
+            const unsigned int n_points = outputs.viscosities.size();
+            
+            if (outputs.template get_additional_output<MaterialModel::AnisotropicViscosity<dim> >() == nullptr)
+            {
+                outputs.additional_outputs.push_back(
+                                                     std::make_shared<MaterialModel::AnisotropicViscosity<dim>> (n_points));

```

Comment:
I *think* that in the current development version, this should be `std_cxx14::make_unique`.